internal/app/app.go

```diff
@@ -0,0 +1,177 @@
+	"strings"
+	"time"
+
+	utiliptables "github.com/gardener/apiserver-proxy/internal/iptables"
+	"github.com/gardener/apiserver-proxy/internal/netif"
+	"k8s.io/klog"
+	"k8s.io/utils/exec"
+)
+
+// NewCacheApp returns a new instance of CacheApp by applying the specified config params.
+func NewCacheApp(params *ConfigParams) (*CacheApp, error) {
+	c := &CacheApp{params: params}
+
+	if params.IPAddress == "" {
+		klog.Infof("No ip-address provided, using environment variable")
+
+		c.localIPStr = os.ExpandEnv("${KUBERNETES_SERVICE_HOST}")
+	} else {
+		klog.Infof("Using IP address %q", params.IPAddress)
+
+		c.localIPStr = params.IPAddress
+	}
+
+	c.localIP = net.ParseIP(c.localIPStr)
+
+	if c.localIP == nil {
+		return nil, fmt.Errorf("unable to lookup IP address of Upstream service kubernetes, ip address was %q", c.localIPStr)
+	}
+
+	klog.Infof("Using %q as master service IP", c.localIPStr)
+
+	return c, nil
+}
+
+// TeardownNetworking removes all custom iptables rules and network interface added by node-cache
+func (c *CacheApp) TeardownNetworking() error {
+	klog.Infof("Cleaning up")
+
+	err := c.netManager.RemoveDummyDevice()
+
+	if c.params.SetupIptables {
+		for _, rule := range c.iptablesRules {
+			exists := true
+			for exists {
+				c.iptables.DeleteRule(rule.table, rule.chain, rule.args...)
+				exists, _ = c.iptables.EnsureRule(utiliptables.Prepend, rule.table, rule.chain, rule.args...)
+			}
+			// Delete the rule one last time since EnsureRule creates the rule if it doesn't exist
+			c.iptables.DeleteRule(rule.table, rule.chain, rule.args...)
+		}
+	}
+
+	return err
+}
+
+func (c *CacheApp) getIPTables() utiliptables.Interface {
+	// using the localIPStr param since we need ip strings here
+	c.iptablesRules = append(c.iptablesRules, []iptablesRule{
+		// Match traffic destined for localIp:localPort and set the flows to be NOTRACKED, this skips connection tracking
+		{utiliptables.Table("raw"), utiliptables.ChainPrerouting, []string{"-p", "tcp", "-d", c.localIPStr,
+			"--dport", c.params.LocalPort, "-j", "NOTRACK"}},
+		// There are rules in filter table to allow tracked connections to be accepted. Since we skipped connection tracking,
+		// need these additional filter table rules.
+		{utiliptables.TableFilter, utiliptables.ChainInput, []string{"-p", "tcp", "-d", c.localIPStr,
+			"--dport", c.params.LocalPort, "-j", "ACCEPT"}},
+		// Match traffic from c.localIPStr:localPort and set the flows to be NOTRACKED, this skips connection tracking
+		{utiliptables.Table("raw"), utiliptables.ChainOutput, []string{"-p", "tcp", "-s", c.localIPStr,
+			"--sport", c.params.LocalPort, "-j", "NOTRACK"}},
+		// Additional filter table rules for traffic frpm c.localIPStr:localPort
+		{utiliptables.TableFilter, utiliptables.ChainOutput, []string{"-p", "tcp", "-s", c.localIPStr,
+			"--sport", c.params.LocalPort, "-j", "ACCEPT"}},
+		// Skip connection tracking for requests to nodelocalDNS that are locally generated, example - by hostNetwork pods
```

Comment:
why is this nodelocal DNS? I see port 443 is used, this is the kubernetes service isn't it?

Reply:
Fixed the comment.